File: macos/Protocol/FrameCodec.swift
import Foundation

public enum Frame {
    case control(Data)
    case binary(fileId: UUID, offset: UInt64, data: Data)
}

public enum FrameCodec {
    public static func encodeControl(_ json: Data) throws -> Data {
        let length = 1 + json.count
        guard length <= ProtocolConstants.maxFrameLength else {
            throw ProtocolError(.protocolViolation, "control frame too large")
        }
        var out = Data()
        out.append(uint32BE(UInt32(length)))
        out.append(ProtocolConstants.kindControl)
        out.append(json)
        return out
    }

    public static func encodeBinary(fileId: UUID, offset: UInt64, data: Data) throws -> Data {
        let length = 1 + 16 + 8 + data.count
        guard length <= ProtocolConstants.maxFrameLength else {
            throw ProtocolError(.protocolViolation, "binary frame too large")
        }
        var out = Data()
        out.append(uint32BE(UInt32(length)))
        out.append(ProtocolConstants.kindBinary)
        out.append(uuidBytes(fileId))
        out.append(uint64BE(offset))
        out.append(data)
        return out
    }

    public static func read(from handle: FileHandle) throws -> Frame {
        let header = try readFully(handle, count: 4)
        let length = Int(getUInt32BE(header))
        if length < 1 || length > ProtocolConstants.maxFrameLength {
            throw ProtocolError(.protocolViolation, "invalid frame length")
        }
        let body = try readFully(handle, count: length)
        let kind = body[0]
        if kind == ProtocolConstants.kindControl {
            return .control(body.dropFirst())
        }
        if kind == ProtocolConstants.kindBinary {
            guard body.count >= 1 + 16 + 8 else {
                throw ProtocolError(.protocolViolation, "short binary frame")
            }
            let fileId = uuidFromBytes(body.subdata(in: 1..<17))
            let offset = getUInt64BE(body.subdata(in: 17..<25))
            let data = body.subdata(in: 25..<body.count)
            return .binary(fileId: fileId, offset: offset, data: data)
        }
        throw ProtocolError(.protocolViolation, "unknown frame kind")
    }

    public static func read(from stream: InputStream) throws -> Frame {
        let header = try readStream(stream, count: 4)
        let length = Int(getUInt32BE(header))
        if length < 1 || length > ProtocolConstants.maxFrameLength {
            throw ProtocolError(.protocolViolation, "invalid frame length")
        }
        let body = try readStream(stream, count: length)
        let kind = body[0]
        if kind == ProtocolConstants.kindControl {
            return .control(body.dropFirst())
        }
        if kind == ProtocolConstants.kindBinary {
            let fileId = uuidFromBytes(body.subdata(in: 1..<17))
            let offset = getUInt64BE(body.subdata(in: 17..<25))
            let data = body.subdata(in: 25..<body.count)
            return .binary(fileId: fileId, offset: offset, data: data)
        }
        throw ProtocolError(.protocolViolation, "unknown frame kind")
    }

    public static func uuidBytes(_ uuid: UUID) -> Data {
        var tuple = uuid.uuid
        return withUnsafeBytes(of: &tuple) { Data($0) }
    }

    public static func uuidFromBytes(_ data: Data) -> UUID {
        var tuple = uuid_t(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)
        _ = withUnsafeMutableBytes(of: &tuple) { ptr in
            data.copyBytes(to: ptr.bindMemory(to: UInt8.self))
        }
        return UUID(uuid: tuple)
    }
}

public func uint32BE(_ value: UInt32) -> Data {
    var be = value.bigEndian
    return withUnsafeBytes(of: &be) { Data($0) }
}

public func uint64BE(_ value: UInt64) -> Data {
    var be = value.bigEndian
    return withUnsafeBytes(of: &be) { Data($0) }
}

public func getUInt32BE(_ data: Data) -> UInt32 {
    let b = [UInt8](data)
    return (UInt32(b[0]) << 24) | (UInt32(b[1]) << 16) | (UInt32(b[2]) << 8) | UInt32(b[3])
}

public func getUInt64BE(_ data: Data) -> UInt64 {
    let b = [UInt8](data)
    var value: UInt64 = 0
    for i in 0..<8 { value = (value << 8) | UInt64(b[i]) }
    return value
}

private func readFully(_ handle: FileHandle, count: Int) throws -> Data {
    var data = Data()
    while data.count < count {
        let chunk = try handle.read(upToCount: count - data.count) ?? Data()
        if chunk.isEmpty { throw ProtocolError(.connectionLost, "eof") }
        data.append(chunk)
    }
    return data
}

private func readStream(_ stream: InputStream, count: Int) throws -> Data {
    var buffer = [UInt8](repeating: 0, count: count)
    var read = 0
    while read < count {
        let n = stream.read(&buffer, maxLength: count - read)
        if n <= 0 { throw ProtocolError(.connectionLost, "eof") }
        read += n
    }
    return Data(buffer)
}

public final class ProtocolSession {
    private let input: InputStream
    private let output: OutputStream
    private let lock = NSLock()

    public init(input: InputStream, output: OutputStream) {
        self.input = input
        self.output = output
    }

    public func send(_ envelope: Envelope) throws {
        let data = try ProtocolJSON.encode(envelope)
        let frame = try FrameCodec.encodeControl(data)
        lock.lock()
        defer { lock.unlock() }
        try writeAll(frame)
    }

    public func sendBinary(fileId: UUID, offset: UInt64, data: Data) throws {
        let frame = try FrameCodec.encodeBinary(fileId: fileId, offset: offset, data: data)
        lock.lock()
        defer { lock.unlock() }
        try writeAll(frame)
    }

    public func receive() throws -> Frame {
        try FrameCodec.read(from: input)
    }

    public func receiveControl() throws -> Envelope {
        switch try receive() {
        case .control(let data):
            return try ProtocolJSON.decode(data)
        case .binary:
            throw ProtocolError(.protocolViolation, "expected control frame")
        }
    }

    private func writeAll(_ data: Data) throws {
        try data.withUnsafeBytes { raw in
            var ptr = raw.bindMemory(to: UInt8.self).baseAddress!
            var remaining = data.count
            while remaining > 0 {
                let n = output.write(ptr, maxLength: remaining)
                if n <= 0 { throw ProtocolError(.connectionLost, "write failed") }
                remaining -= n
                ptr += n
            }
        }
        output.outputStreamWriteFlush()
    }
}

private extension OutputStream {
    func outputStreamWriteFlush() {
        // OutputStream has no flush; TCP send buffer is enough.
    }
}

public func makeEnvelope(type: MessageType, payload: [String: JSONValue], msgId: String = UUID().uuidString, ts: Int64 = Int64(Date().timeIntervalSince1970 * 1000)) -> Envelope {
    Envelope(type: type.rawValue, msgId: msgId, ts: ts, payload: payload)
}

File: android/protocol/src/main/kotlin/com/honor/share/protocol/FrameCodec.kt
package com.honor.share.protocol

import java.io.EOFException
import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.UUID

sealed class Frame {
    data class Control(val json: ByteArray) : Frame() {
        override fun equals(other: Any?): Boolean =
            other is Control && json.contentEquals(other.json)

        override fun hashCode(): Int = json.contentHashCode()
    }

    data class Binary(
        val fileId: UUID,
        val offset: Long,
        val data: ByteArray,
    ) : Frame() {
        override fun equals(other: Any?): Boolean =
            other is Binary && fileId == other.fileId && offset == other.offset && data.contentEquals(other.data)

        override fun hashCode(): Int {
            var result = fileId.hashCode()
            result = 31 * result + offset.hashCode()
            result = 31 * result + data.contentHashCode()
            return result
        }
    }
}

object FrameCodec {
    fun encodeControl(json: ByteArray): ByteArray {
        val length = 1 + json.size
        require(length <= ProtocolConstants.MAX_FRAME_LENGTH) { "control frame too large" }
        val out = ByteArray(4 + length)
        putIntBe(out, 0, length)
        out[4] = ProtocolConstants.KIND_CONTROL.toByte()
        json.copyInto(out, 5)
        return out
    }

    fun encodeBinary(fileId: UUID, offset: Long, data: ByteArray): ByteArray {
        val payload = 1 + 16 + 8 + data.size
        require(payload <= ProtocolConstants.MAX_FRAME_LENGTH) { "binary frame too large" }
        val out = ByteArray(4 + payload)
        putIntBe(out, 0, payload)
        out[4] = ProtocolConstants.KIND_BINARY.toByte()
        uuidToBytes(fileId).copyInto(out, 5)
        putLongBe(out, 21, offset)
        data.copyInto(out, 29)
        return out
    }

    fun write(output: OutputStream, frame: ByteArray) {
        output.write(frame)
        output.flush()
    }

    fun read(input: InputStream): Frame {
        val header = readFully(input, 4)
        val length = getIntBe(header, 0)
        if (length < 1 || length > ProtocolConstants.MAX_FRAME_LENGTH) {
            throw ProtocolException(ErrorCode.PROTOCOL_VIOLATION, "invalid frame length")
        }
        val body = readFully(input, length)
        return when (body[0].toInt() and 0xFF) {
            ProtocolConstants.KIND_CONTROL -> Frame.Control(body.copyOfRange(1, body.size))
            ProtocolConstants.KIND_BINARY -> {
                if (body.size < 1 + 16 + 8) {
                    throw ProtocolException(ErrorCode.PROTOCOL_VIOLATION, "short binary frame")
                }
                val fileId = bytesToUuid(body, 1)
                val offset = getLongBe(body, 17)
                val data = body.copyOfRange(25, body.size)
                Frame.Binary(fileId, offset, data)
            }
            else -> throw ProtocolException(ErrorCode.PROTOCOL_VIOLATION, "unknown frame kind")
        }
    }

    fun uuidToBytes(id: UUID): ByteArray {
        val buffer = ByteBuffer.allocate(16).order(ByteOrder.BIG_ENDIAN)
        buffer.putLong(id.mostSignificantBits)
        buffer.putLong(id.leastSignificantBits)
        return buffer.array()
    }

    fun bytesToUuid(bytes: ByteArray, offset: Int = 0): UUID {
        val buffer = ByteBuffer.wrap(bytes, offset, 16).order(ByteOrder.BIG_ENDIAN)
        return UUID(buffer.long, buffer.long)
    }

    private fun putIntBe(target: ByteArray, offset: Int, value: Int) {
        target[offset] = (value ushr 24).toByte()
        target[offset + 1] = (value ushr 16).toByte()
        target[offset + 2] = (value ushr 8).toByte()
        target[offset + 3] = value.toByte()
    }

    private fun putLongBe(target: ByteArray, offset: Int, value: Long) {
        for (i in 0 until 8) {
            target[offset + i] = (value ushr ((7 - i) * 8)).toByte()
        }
    }

    private fun getIntBe(source: ByteArray, offset: Int): Int {
        return ((source[offset].toInt() and 0xFF) shl 24) or
            ((source[offset + 1].toInt() and 0xFF) shl 16) or
            ((source[offset + 2].toInt() and 0xFF) shl 8) or
            (source[offset + 3].toInt() and 0xFF)
    }

    private fun getLongBe(source: ByteArray, offset: Int): Long {
        var value = 0L
        for (i in 0 until 8) {
            value = (value shl 8) or (source[offset + i].toLong() and 0xFF)
        }
        return value
    }

    private fun readFully(input: InputStream, count: Int): ByteArray {
        val buffer = ByteArray(count)
        var read = 0
        while (read < count) {
            val n = input.read(buffer, read, count - read)
            if (n < 0) throw EOFException("unexpected end of stream")
            read += n
        }
        return buffer
    }
}

class ProtocolException(
    val code: ErrorCode,
    message: String,
    cause: Throwable? = null,
) : Exception(message, cause)
